### PR TITLE
[IMP] l10n_in: deprecate multiple GSTN units in a single company.

### PR DESCRIPTION
**Pull request by jva-odoo**
Deprecate the implementation of multiple GSTN units using journal id on sales, purchase and warehouse. Use multi-company incase you need the multiple GSTN.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jco-odoo** [comment]
@jva-odoo Did you think about migration scripts for this?

File: addons/l10n_in/models/account.py [hunk 1]
```diff
@@ -10,23 +10,7 @@ class AccountJournal(models.Model):
...
-        result_dict = dict(result)
-        indian_journals = self.filtered(lambda j: j.company_id.account_fiscal_country_id.code == 'IN' and
-            j.l10n_in_gstin_partner_id and j.l10n_in_gstin_partner_id.vat)
-        for journal in indian_journals:
-            name = result_dict[journal.id]
-            name += "- %s" % (journal.l10n_in_gstin_partner_id.vat)
-            result_dict[journal.id] = name
-        return list(result_dict.items())
-
+    l10n_in_gstin_partner_id = fields.Many2one('res.partner', string="GSTIN Unit (Deprecated)", ondelete="restrict", help="GSTIN related to this journal. If empty then consider as company GSTIN.", deprecated=True)
```
Comment (jco-odoo) [inline addons/l10n_in/models/account.py:13]: Why do you keep the field in deprecated True?  (why not remove it?)

File: addons/l10n_in/models/account_invoice.py [hunk 2]
```diff
@@ -92,7 +92,7 @@ def _post(self, soft=True):
                     move.l10n_in_state_id = self._l10n_in_get_indian_state(move.partner_id)
                 #still state is not set then assumed that transaction is local like PoS so set state of company unit
                 if not move.l10n_in_state_id:
-                    move.l10n_in_state_id = company_unit_partner.state_id
+                    move.l10n_in_state_id = company.state_id
```
Comment (jco-odoo) [inline addons/l10n_in/models/account_invoice.py:95]: It is weird to still keep the l10n_in_state_id if the only reason is the l10n_in.state_in_ot

File: addons/l10n_in_purchase/models/purchase_order.py [hunk 3]
```diff
@@ -8,8 +8,9 @@
 class PurchaseOrder(models.Model):
     _inherit = "purchase.order"
 
-    l10n_in_journal_id = fields.Many2one('account.journal', string="Journal", \
-        states=Purchase.READONLY_STATES, domain="[('type', '=', 'purchase')]")
+    #Deprecated field, use multi-company instead of this field
+    l10n_in_journal_id = fields.Many2one('account.journal', string="Journal (Deprecated)", \
+        states=Purchase.READONLY_STATES, domain="[('type', '=', 'purchase')]", deprecated=True)
```
Comment (jco-odoo) [inline addons/l10n_in_purchase/models/purchase_order.py:13]: same here: let's remove it then...

File: addons/l10n_in_sale_stock/models/stock_warehouse.py [hunk 4]
```diff
@@ -7,4 +7,5 @@
 class Stock(models.Model):
     _inherit = 'stock.warehouse'
 
-    l10n_in_sale_journal_id = fields.Many2one('account.journal', string="Sale Journal")
+    #Deprecated field, use multi-company instead of this field
+    l10n_in_sale_journal_id = fields.Many2one('account.journal', string="Sale Journal(Deprecated)", deprecated=True)
```
Comment (jco-odoo) [inline addons/l10n_in_sale_stock/models/stock_warehouse.py:11]: same

File: addons/l10n_in_sale/models/sale_order.py [hunk 5]
```diff
@@ -9,7 +9,8 @@ class SaleOrder(models.Model):
 
     l10n_in_reseller_partner_id = fields.Many2one('res.partner',
         string='Reseller', domain="[('vat', '!=', False), '|', ('company_id', '=', False), ('company_id', '=', company_id)]", readonly=True, states={'draft': [('readonly', False)], 'sent': [('readonly', False)]})
-    l10n_in_journal_id = fields.Many2one('account.journal', string="Journal", compute="_compute_l10n_in_journal_id", store=True, readonly=True, states={'draft': [('readonly', False)], 'sent': [('readonly', False)]})
```
Comment (jco-odoo) [inline addons/l10n_in_sale/models/sale_order.py:12]: same

**Review by jco-odoo — COMMENTED** [review]
@mga-odoo @jva-odoo Either we keep it or we remove it, but if we remove it, we have to do it entirely and provide the necessary upgrade scripts.

File: addons/l10n_in/models/account_invoice.py (hunk 2 above)
Comment (jva-odoo) [reply addons/l10n_in/models/account_invoice.py:95]: @jco-odoo l10n_in_state_id is not for this only but this is Place of Supply and there is another PR of this where we not remove it but simplify it.

File: addons/l10n_in/models/account.py (hunk 1 above)
Comment (jva-odoo) [reply addons/l10n_in/models/account.py:13]: @jco-odoo If we directly remove it then we need to create a company and move all data to the newly created company.

And this is super complex to move all data to new companies that link to that unit to like invoices>sales->subscriptions. stock that related to sales etc.

Comment (jco-odoo) [reply addons/l10n_in/models/account.py:13]: @jva-odoo But it is what needs to happen.  Complex or not.

Comment (jva-odoo) [reply addons/l10n_in/models/account.py:13]: @jco-odoo Okay, can we just add deprecated and remove it in the next version?
so we have big-time to write complex migration.

**Comment by jva-odoo** [comment]
@jco-odoo we look at this when related migration script is ready.

**Comment by jva-odoo** [comment]
move to https://github.com/odoo/odoo/pull/107234

**Pull request closed without merging**